bundles/org.openhab.binding.hive/src/main/java/org/openhab/binding/hive/internal/client/FeatureAttributeFactory.java

```diff
@@ -0,0 +1,122 @@
+
+    private FeatureAttributeFactory() {
+        throw new AssertionError();
+    }
+
+    private static <F, T> void buildFeatureAttribute(
+            final DefaultFeatureAttribute.Builder<T> featureAttributeBuilder,
+            final Function<F, T> adapter,
+            final FeatureAttributeDto<F> dto
+    ) {
+        Objects.requireNonNull(featureAttributeBuilder);
+        Objects.requireNonNull(adapter);
+        Objects.requireNonNull(dto);
+
+        final @Nullable F displayValue = dto.displayValue;
+        final @Nullable F reportedValue = dto.reportedValue;
+        final @Nullable HiveApiInstant reportChangedTime = dto.reportChangedTime;
+        final @Nullable HiveApiInstant reportReceivedTime = dto.reportReceivedTime;
+
+        if (displayValue == null) {
+            throw new HiveClientResponseException("Display value is unexpectedly null.");
+        }
+        featureAttributeBuilder.displayValue(adapter.apply(displayValue));
+
+        if (reportedValue == null) {
+            throw new HiveClientResponseException("Reported value is unexpectedly null.");
+        }
+        featureAttributeBuilder.reportedValue(adapter.apply(reportedValue));
+
+        if (reportChangedTime == null) {
+            throw new HiveClientResponseException("Report Changed Time is unexpectedly null.");
+        }
+        featureAttributeBuilder.reportChangedTime(reportChangedTime.asInstant());
+
+        if (reportReceivedTime == null) {
+            throw new HiveClientResponseException("Reported Received Time is unexpectedly null.");
+        }
+        featureAttributeBuilder.reportReceivedTime(reportReceivedTime.asInstant());
+    }
+
+    public static <T> FeatureAttribute<T> getReadOnlyFromDto(final @Nullable FeatureAttributeDto<T> dto) {
+        return getReadOnlyFromDtoWithAdapter(Function.identity(), dto);
+    }
+
+    public static <F, T> FeatureAttribute<T> getReadOnlyFromDtoWithAdapter(
+            final Function<F, T> adapter,
+            final @Nullable FeatureAttributeDto<F> dto
+    ) {
+        if (dto == null) {
+            throw new HiveClientResponseException(FEATURE_ATTRIBUTE_NULL_MESSAGE);
+        }
+
+        final DefaultFeatureAttribute.Builder<T> featureAttributeBuilder = DefaultFeatureAttribute.builder();
+
+        buildFeatureAttribute(
+                featureAttributeBuilder,
+                adapter,
+                dto
+        );
+
+        return featureAttributeBuilder.build();
+    }
+
+    public static <T> SettableFeatureAttribute<T> getSettableFromDto(final @Nullable FeatureAttributeDto<T> dto) {
+        return getSettableFromDtoWithAdapter(Function.identity(), dto);
+    }
+
+    public static <F, T> SettableFeatureAttribute<T> getSettableFromDtoWithAdapter(
+            final Function<F, T> adapter,
+            final @Nullable FeatureAttributeDto<F> dto
+    ) {
+        if (dto == null) {
+            throw new HiveClientResponseException(FEATURE_ATTRIBUTE_NULL_MESSAGE);
+        }
```

Comment:
The idea was in instead of doing lots of null checks in the `DefaultNodeRepository#getXyzFeatureFromDto(...)` methods I could do the checks in one place.  This meant I had to make the parameter `@Nullable` so the static analysis would be happy.